keras_nlp/layers/modeling/rotary_embedding.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.backend import keras
+from keras_nlp.backend import ops
+
+
+@keras_nlp_export("keras_nlp.layers.RotaryEmbedding")
+class RotaryEmbedding(keras.layers.Layer):
+    """Rotary positional encoding layer.
+
+    This layer encodes absolute positional information with rotation
+    matrix. It calculates the rotary encoding with a mix of sine and
+    cosine functions with geometrically increasing wavelengths.
+    Defined and formulated in [RoFormer: Enhanced Transformer with Rotary Position Embedding](https://arxiv.org/abs/2104.09864v4).
+    Takes as input the query and key tensors. The input must have shape
+    [batch_size, num_heads, sequence_length, query_length]. This layer
+    will return new query and key tensors after applying rotational
+    encoding.
+
+    Args:
+        percentage: float. The percentage of attn_head_size over which rotation
+            should be applied. Defaults to 0
+        max_wavelength: int. The maximum angular wavelength of the sine/cosine
+            curves, as described in Attention is All You Need. Defaults to
+            `10000`.
+
+    Examples:
+
+    ```python
```

Comment:
We should include two examples here, one with `(batch_size, sequence_length, feature_length)` inputs, and one with `(batch_size, sequence_length, num_heads, feature_length)` inputs.